### PR TITLE
Handle more cases of typos misinterpreted as type ascription

### LINKED ISSUE #60933: Type ascription typo yields unclear error message
The following code does not guide the user towards the problem. Especially considering the fact that there's no type ascription feature gate active.

```rust
fn main() {
    let u: usize = std::mem:size_of::<u32>();
    //                     ^ note the single colon instead of double
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=17a23068f90f2e62cdf7fde4c5592953))

Errors:

```
   Compiling playground v0.0.1 (/playground)
warning: unnecessary path disambiguator
 --> src/main.rs:2:36
  |
2 |     let u: usize = std::mem:size_of::<u32>();
  |                                    ^^ try removing `::`

error: expected one of `!`, `::`, or `;`, found `(`
 --> src/main.rs:2:43
  |
2 |     let u: usize = std::mem:size_of::<u32>();
  |                                           ^ expected one of `!`, `::`, or `;` here

error: aborting due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

### PR DESCRIPTION
**Pull request by estebank**
Fix #60933, #54516.

CC #47666, #34255, #48016.

### CODE REVIEW

File: src/test/ui/suggestions/type-ascription-instead-of-variant.stderr [hunk 1]
```diff
@@ -2,9 +2,12 @@ error: expected type, found `""`
...
    |
 LL |     let _ = Option:Some("");
-   |                   -     ^^ expecting a type here because of type ascription
+   |                   -     ^^ expected type
    |                   |
-   |                   help: maybe you meant to write a path separator here: `::`
+   |                   help: maybe write a path separator here: `::`
+   |
+   = note: `#![feature(type_ascription)]` lets you annotate an expression with a type: `<expr>: <type>` 3
+   = note: for more information, see https://github.com/rust-lang/rust/issues/23416
```
Comment (CryZe) [inline src/test/ui/suggestions/type-ascription-instead-of-variant.stderr]: What's the trailing 3 for?

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -2840,9 +2846,13 @@ impl<'a> Parser<'a> {
                 self.parse_prefix_expr(attrs)?
             }
         };
+        let last_type_ascription_set = self.last_type_ascription.is_some();
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Unnecessary condition: in all cases below `self.last_type_ascription = None;` can be done unconditionally.

File: src/libsyntax/parse/diagnostics.rs [hunk 3]
```diff
@@ -365,9 +366,53 @@ impl<'a> Parser<'a> {
...
         Err(err)
     }
 
+    pub fn maybe_annotate_with_ascription(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        maybe_expected_semicolon: bool,
+    ) {
+        if let Some((sp, likely_path)) = self.last_type_ascription {
+            let cm = self.sess.source_map();
```
Comment (petrochenkov) [inline src/libsyntax/parse/diagnostics.rs]:
```suggestion
            let sm = self.sess.source_map();
```
Nit: remains of `code_map`.

**Comment by petrochenkov** [comment]
r=me with nits addressed and the debugging `3` removed

File: src/libsyntax/parse/parser.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: The condition is necessary so that 3027 doesn't erase the assignment in 2940. All other cases don't need it.

**Comment by frehberg** [comment]
I am facing a similar issue, but just when using "cargo bench"

```rust
   
    #[bench_resources("res/*/input.txt")]
    fn measure_resource(b: &mut test::Bencher, resource: &str) {
        let path = std::path::Path::new(resource);
        b.iter(|| path.exists());
    }
```
yields the following error


```
frehberg@mochila:~/src/proj-dds/test-generator/example$ cargo bench
   Compiling test-generator-example v0.3.0 (/home/frehberg/src/proj-dds/test-generator/example)
error[E0658]: type ascription is experimental
  --> example/benches/mybenches.rs:30:5
   |
30 |     #[bench_resources("res/*/input.txt")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: for more information, see https://github.com/rust-lang/rust/issues/23416
   = help: add `#![feature(type_ascription)]` to the crate attributes to enable

error: aborting due to previous error
```

I did an update to nightly "rustc 1.38.0-nightly (c798dffac 2019-07-27)", but issue still exists

any idea?

**Comment by frehberg** [comment]
Sorry, I removed my comment, it was caused by an error in my code.

**Pull request merged**